Codec/Archive/Tar/Check/Internal.hs

```diff
@@ -33,14 +37,23 @@ module Codec.Archive.Tar.Check.Internal (
 import Codec.Archive.Tar.Types
 import Control.Applicative ((<|>))
 import qualified Data.ByteString.Lazy.Char8 as Char8
+import Data.Maybe (fromMaybe)
 import Data.Typeable (Typeable)
-import Control.Exception (Exception)
+import Control.Exception (Exception(..))
 import qualified System.FilePath as FilePath.Native
-         ( splitDirectories, isAbsolute, isValid, (</>), takeDirectory )
+         ( splitDirectories, isAbsolute, isValid, (</>), takeDirectory, hasDrive )
 
 import qualified System.FilePath.Windows as FilePath.Windows
 import qualified System.FilePath.Posix   as FilePath.Posix
 
+instance forall e1 e2 . (Exception e1, Exception e2) => Exception (Either e1 e2) where
+  toException (Left e)  = toException e
+  toException (Right e) = toException e
+  fromException se = case fromException @e1 se of
+                       Nothing -> Right <$> fromException @e2 se
+                       Just e  -> Just (Left e)
+  displayException (Left e)  = displayException e
+  displayException (Right e) = displayException e
```

Comment:
Is this evil? But otherwise we can't really combine `unpackRaw "foo" (checkPortability. checkSecurity $ entries)`

Reply:
I'm also not sure how well `checkPortability . checkSecurity` fuses or if it does at all.

I've been thinking maybe it's a better API to provide a callback for `unpack` that checks entries.

Reply:
> I'm also not sure how well `checkPortability . checkSecurity` fuses or if it does at all.

It does not and never did, there are no fusion rules for `mapEntries`. This is not a concern for now; if we ever care about performance, avoiding `String` would be a much bigger fish to fry.

> Is this evil?

Orphan instance **are** evil, I'm afraid.

Other than the instance, looks good to me.

Reply:
Well, how do you implement this without the instance?

Reply:
How about 
```haskell
unpackRaw :: (e -> IO ()) -> FilePath -> Entries e -> IO ()
```
?

Reply:
I designed 'checkSecurity' and friends as callbacks.